DOC: Improve the docstring of pd.Index.contains and closes PR #20211

TanyaaCJain opened the pull request:
- [x] closes #20211 
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

This PR is an update to PR #20211, improving the `Index.contains` docstring.

pep8speaks commented:
Hello @Tanya-Jain! Thanks for submitting the PR.

 - There are no PEP8 issues in the file [`pandas/core/indexes/base.py`](https://github.com/Tanya-Jain/pandas/blob/1d83db4a0965726f94b885a103073dd4ee4f3fc8/pandas/core/indexes/base.py) !

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -2005,15 +2005,37 @@ def __contains__(self, key):
+
+        See Also
+        --------
+        CategoricalIndex : Returns a bool if the 1-dimensional, Categorical key
```
`CategoricalIndex` is the class, do you mean `CategoricalIndex.contains` here?

Tanya-Jain replied:
Yes, I mean this. Apparently, there isn't any docstring written for the same. Should I write something like 
`CategoricalIndex : Its :method: contains returns a bool if the key is in index. The key can be of the type same as this class, like the Index.contains` 
or should I still refer to `CategoricalIndex.contains : Returns a bool if the key is in index . . .`

datapythonista replied:
The `Index` class is a parent of `CategoricalIndex` as well as all the other index classes like `Float64Index`... It could make sense to list `CategoricalIndex.contains`, `Float64Index.contains`... but I don't think it adds that much value in this case, and it adds noise to the docstring.

I think the best is just leave `Index.isin`, but it's up to you.

Tanya-Jain replied:
I agree as `CategoricalIndex.contains` behaves similarly to `Index.contains`. Let's just keep `Index.isin`.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -2005,15 +2005,37 @@ def __contains__(self, key):
+
+        Examples
+        --------
+        >>> l1 = pd.Index([1, 2, (3, 4), 5])
```
By convention we use `idx` for variables soring an index in the examples (`df` for dataframes, and `s` for series). Can you use that?

Also, can you show the content of the index after you store it (i.e. `>>> idx`)

Tanya-Jain replied:
Yes, I can change the variable naming and show the content, it would be better.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -2005,15 +2005,37 @@ def __contains__(self, key):
+        >>> l1 = pd.Index([1, 2, (3, 4), 5])
+        >>> t = (3, 4)
+        >>> num1 = 1
+        >>> num2 = 6
```
I'd save a bit of space by using the values directly in the calls, instead of saving variables first (e.g. `idx.contains(1)`

Tanya-Jain replied:
I will improve this.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -2005,15 +2005,37 @@ def __contains__(self, key):
         ----------
-        key : object
+        key : label
+            The key requested. Immutable-like, 1-dimensional.
```
While true, I think users will find this a bit confusing. I think it probably makes more sense to avoid saying that is 1-dimensional, as it gives the impression that more than one key can be provided.

Personally, I think the users should look for what a label can be (immutable, 1-dimensional if a tuple...) in some other part of the documentation, and here focus more on what `contains` exactly expects (the key that will be searched in the index). But if you prefer to be specific, I'd start this by something like `The key can be of the same type as the labels in index, so it needs to be immutable...`.

Tanya-Jain replied:
This is exactly what I tried to explain, thank you!

datapythonista reviewed (commented):
Good work, I added some comments about minor things.

datapythonista commented:
Any reason to close this? I guess it's been by mistake?

Tanya-Jain commented:
#23100 contained an additional commit for merging commits from the pandas-dev/pandas master. 
To have only the commit for the file changes made in base.py, I closed #23100 and opened #23107

TanyaaCJain committed:
DOC: Improve the docstring of pd.Index.contains

Tanya-Jain commented:
I tried to get the problem fixed but ended up having 0 commits being shown as well as having this PR automatically closed.

TanyaaCJain committed:
DOC: Improvement on Index.contains for reviews on PR #23100

- Description for key has more clarity on its behaviour.
- Removed :class: `CategoricalIndex` from the See Also section as
  the methods: CategoricalIndex.contains, Float64Index.contains and
  like, behave similarly to `Index.contains`. Hence, keeping only
  `Index.isin`
- Use `idx` for naming variables for objects of :class: `Index`
  instead of `l` for passing list-like key, to encourage pandas
  docstring standards and clarity.
- Directly call the values in :method: `Index.contains` to prevent
  additional memory usage.

TanyaaCJain committed:
Merge branch 'master' of https://github.com/pandas-dev/pandas into pr-update

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -2005,15 +2005,35 @@ def __contains__(self, key):
 
     _index_shared_docs['contains'] = """
-        return a boolean if this key is IN the index
+        Return a boolean if this key is in the index.
```
Minor thing, and I know this is the original description and not yours, but this sounds like it a boolean is returned only if the key is in the index. Not that a boolean is always returned, on whether the key is in the index. Do you mind rephrasing it to be more correct please?

Tanya-Jain replied:
Sure, I'll rephrase it in the next commit as a boolean indicator. Apparently, there are few more docstrings with similar phrasing like [`Series.is_unique`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.is_unique.html#pandas.Series.is_unique), [`Series.is_monotonic`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.is_monotonic.html#pandas.Series.is_monotonic), [`Series.is_monotonic_increasing`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.is_monotonic_increasing.html#pandas.Series.is_monotonic_increasing), [`Series.is_monotonic_decreasing`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.is_monotonic_decreasing.html#pandas.Series.is_monotonic_decreasing). These too can be corrected in future if seems appropriate.

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -2005,15 +2005,35 @@ def __contains__(self, key):
         ----------
-        key : object
+        key : label
+            The key can be of the same type as the label of :class: `Index`,
```
I think the space after `:class:` is not needed. Did you check if this is rendered correctly in the html version? You should be able to generate this page with `./doc/make.py html --single=pandas.Index.contains`

Tanya-Jain replied:
Thanks, it renders correctly after removing the space after `:class:`

datapythonista commented on pandas/core/indexes/base.py:
```diff
@@ -2005,15 +2005,35 @@ def __contains__(self, key):
+        >>> idx
+        Index([1, 2, (3, 4), 5], dtype='object')
+        >>> idx.contains((3,4))
+        True
```
Small thing, but do you mind moving this example after the other two? I think it helps if we show the simple/easy cases first, and the more complex later.

Also, you there is a missing space after the comma.

When you're done with the changes, it may be helpful to run `./scripts/validate_docstrings.py pandas.Index.contains` to make sure there are not errors (I don't think there are).

And also `flake8 --doctests pandas/core/indexes/base.py`. Note that this will report all doctests problems in all the docstrings of the file. Just look that there is no problem in the lines of this docstring 2008-2036.

Tanya-Jain replied:
I have made the said changes.

Validation tests passed with a warning of absence of an extended description which has been suggested to avoid in the review of the original PR. Tested via command `./scripts/validate_docstrings.py pandas.Index.contains`

No doctests problem were reported for lines 2008-2036 by testing via command `flake8 --doctests pandas/core/indexes/base.py`

datapythonista reviewed (commented):
It looks very good. Just couple of small things, and it's ready to be merged.

TanyaaCJain committed:
Rephrasing, Testing and Validating docstring

- Rephrased description as a boolean indicator
- Removed a space after :class: for correct rendering checked by
  the command `./doc/make.py html --single=pandas.Index.contains`
- Moved the easier example cases first.
- Added the missing space after comma following PEP8
- Validation tests passed with a warning of absence of an extended
  description. Tested via command `./scripts/validate_docstrings.py
  pandas.Index.contains`
- No doctests problem were reported for lines 2008-2036 by testing
  via command `flake8 --doctests pandas/core/indexes/base.py`

TanyaaCJain committed:
Merge remote-tracking branch 'upstream/master' into pr-update

Tanya-Jain commented:
- [x] The html page generates correctly with `./doc/make.py html --single=pandas.Index.contains`
- [x] Validation tests passed with a warning of absence of an extended description as asked in the #20211 PR review. Tested via command `./scripts/validate_docstrings.py pandas.Index.contains`
- [x] No doctests problem were reported for lines 2008-2036 by testing via command `flake8 --doctests pandas/core/indexes/base.py`

Validation Script:
```
################################################################################
###################### Docstring (pandas.Index.contains)  ######################
################################################################################

Return a boolean indicating whether this key is in the index.

Parameters
----------
key : label
    The key can be of the same type as the label of :class:`Index`,
    hence immutable-like and 1-dimensional if it is a tuple.

Returns
-------
bool
    Result indicating whether the key search is in the index.

See Also
--------
Index.isin : Returns an ndarray of boolean dtype indicating whether the
    list-like key is in the index.

Examples
--------
>>> idx = pd.Index([1, 2, (3, 4), 5])
>>> idx
Index([1, 2, (3, 4), 5], dtype='object')
>>> idx.contains(1)
True
>>> idx.contains(6)
False
>>> idx.contains((3, 4))
True

################################################################################
################################## Validation ##################################
################################################################################

Warnings found:
	No extended summary found
Docstring for "pandas.Index.contains" correct. :)
```

datapythonista reviewed (approved):
lgtm. Thanks @Tanya-Jain

datapythonista commented:
@jreback if you don't mind reviewing and merging this one too

jreback commented on pandas/core/indexes/base.py:
```diff
@@ -2019,15 +2019,35 @@ def __contains__(self, key):
+
+        Examples
+        --------
+        >>> idx = pd.Index([1, 2, (3, 4), 5])
```
I would prefer an example with a simple values like ``pd.Index(list('abcd'))``. This is a rather esoteric case.

datapythonista commented:
@Tanya-Jain do you have time to address the comments of the last review and fix the conflicts?

datapythonista committed:
Merging from master

datapythonista committed:
Improving Index.contains example, and reusing docstring for __contains__

datapythonista committed:
Improving texts of the docstring

datapythonista commented:
@jreback addressed your comments, and made some other fixes to this docstring, if you can take a look.

jreback commented:
thanks @Tanya-Jain and @datapythonista
